cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/RuntimeJobTwillController.java

```diff
@@ -16,222 +16,66 @@
-  }
-
-  @Override
-  public RunId getRunId() {
-    return runId;
-  }
-
-  @Override
-  public Future<Command> sendCommand(Command command) {
-    return Futures.immediateFailedFuture(new UnsupportedOperationException("sendCommand is not supported"));
-  }
-
-  @Override
-  public Future<Command> sendCommand(String runnableName, Command command) {
-    return Futures.immediateFailedFuture(new UnsupportedOperationException("sendCommand is not supported"));
-  }
-
-  @Override
-  public void onRunning(Runnable runnable, Executor executor) {
-    started.thenRunAsync(runnable, executor);
-  }
-
-  @Override
-  public void onTerminated(Runnable runnable, Executor executor) {
-    completion.whenCompleteAsync((remoteExecutionTwillController, throwable) -> runnable.run(), executor);
-  }
-
-  @Override
-  public void awaitTerminated() throws ExecutionException {
-    Uninterruptibles.getUninterruptibly(completion);
-  }
-
-  @Override
-  public void awaitTerminated(long timeout, TimeUnit timeoutUnit) throws TimeoutException, ExecutionException {
-    Uninterruptibles.getUninterruptibly(completion, timeout, timeoutUnit);
-  }
-
   @Override
   public Future<? extends ServiceController> terminate() {
-    if (terminateCalled.compareAndSet(false, true)) {
-      try {
-        // stop the job
-        jobManager.stop(new ProgramRunInfo.Builder()
-                          .setNamespace(programRunId.getNamespace())
-                          .setApplication(programRunId.getApplication())
-                          .setProgram(programRunId.getProgram())
-                          .setProgramType(programRunId.getType().getPrettyName())
-                          .setRun(programRunId.getRun()).build());
-        // mark completion as completed
-        completion.complete(this);
-      } catch (Exception e) {
-        completion.completeExceptionally(e);
-      }
+    if (getTerminationFuture().isDone()) {
+      return CompletableFuture.completedFuture(this);
     }
-    return completion;
-  }
 
-  @Override
-  public void kill() {
-    if (terminateCalled.compareAndSet(false, true)) {
+    CompletableFuture<TwillController> result = getTerminationFuture().thenApply(r -> r);
+    executor.execute(() -> {
       try {
-        // stop the job
-        jobManager.kill(new ProgramRunInfo.Builder()
-                          .setNamespace(programRunId.getNamespace())
-                          .setApplication(programRunId.getApplication())
-                          .setProgram(programRunId.getProgram())
-                          .setProgramType(programRunId.getType().getPrettyName())
-                          .setRun(programRunId.getRun()).build());
-        // mark completion as completed
-        completion.complete(this);
+        if (jobManager.getDetail(programRunInfo).isPresent()) {
+          // stop the job
+          jobManager.stop(programRunInfo);
```

Comment:
who is responsible for calling `jobManager.kill`?

Reply:
oh right. Missed that during refactoring